trustfall_core/src/interpreter/basic_adapter.rs

```diff
@@ -142,6 +149,48 @@ pub trait BasicAdapter<'vertex> {
         type_name: &str,
         coerce_to_type: &str,
     ) -> ContextOutcomeIterator<'vertex, Self::Vertex, bool>;
+
+    /// Resolve the `__typename` special property over an iterator of query contexts.
+    ///
+    /// Each [`DataContext`] in the `contexts` argument has an active vertex,
+    /// which is either `None`, or a `Some(Self::Vertex)` value representing a vertex
+    /// of type `type_name` defined in the schema.
+    ///
+    /// This method resolves the name of the type of that active vertex.
+    ///
+    /// The default implementation uses the [`Typename`] trait implemented by `Self::Vertex`
+    /// to get each vertex's type name.
+    ///
+    /// The caller guarantees that:
+    /// - `type_name` is a type or interface defined in the schema.
+    /// - When the active vertex is `Some(...)`, it's a vertex of type `type_name`:
+    ///   either its type is exactly `type_name`, or `type_name` is an interface that
+    ///   the vertex's type implements.
+    ///
+    /// The returned iterator must satisfy these properties:
+    /// - Produce `(context, property_value)` tuples with the property's value for that context.
+    /// - Produce contexts in the same order as the input `contexts` iterator produced them.
+    /// - Produce property values whose type matches the property's type defined in the schema.
+    /// - When a context's active vertex is `None`, its property value is `FieldValue::Null`.
+    ///
+    /// # Overriding the default implementation
+    ///
+    /// Some adapters may be able to implement this method more efficiently than the provided
+    /// default implementation.
+    ///
+    /// For example: adapters having access to a [`Schema`] can use
+    /// the [`interpreter::helpers::resolve_typename`](super::helpers::resolve_typename) method,
+    /// which implements a "fast path" for types that don't have any subtypes per the schema.
+    ///
+    /// [`DataContext`]: super::DataContext
+    /// [`Schema`]: crate::schema::Schema
+    fn resolve_typename(
+        &mut self,
+        contexts: ContextIterator<'vertex, Self::Vertex>,
+        _type_name: &str,
+    ) -> ContextOutcomeIterator<'vertex, Self::Vertex, FieldValue> {
+        resolve_property_with(contexts, |vertex| vertex.typename().into())
+    }
```

Comment:
@ginger51011 curious what you think about this design choice: `BasicAdapter` handles typenames automatically but allows that logic to be overridden as well.

Reply:
I think it's very reasonable, but what is the use of the `_type_name` parameter right now? A quick glance at the docs doesn't explain it and it is a bit confusing to have `resolve_typename` require a typename

Reply:
Good point, as always! I expanded that bit of the docs with a small example. Let me know what you think!

Reply:
Yes, that makes more sense

Reply:
Awesome! Thanks for checking it out, much appreciated!